### PR TITLE
Hebmorph 2.1

### PR DESCRIPTION
**Pull request by EgozyN**

### CODE REVIEW

**Comment by synhershko** [comment]
This change warrants a version 3.0 release, not a 2.1

**Comment by synhershko** [comment]
Please add tests for each new TokenFilter implementation added

File: java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java [hunk 1]
```diff
@@ -39,40 +44,17 @@ public HebrewQueryLightAnalyzer() throws IOException {
...
-                duplicateCurrentToken();
-                suffixCurrent();
-            }
-        };
+        TokenStream tok = new NiqqudFilter(src);
+        tok = new ASCIIFoldingFilter(tok);
+        tok = new LowerCaseFilter(tok);
+        tok = new HebrewLemmatizerTokenFilter(tok, dict, false, false);
+        tok = new IgnoreOriginalTokenFilter(tok);
+        tok = new AddSuffixTokenFilter(tok, '$');
```
Comment (synhershko) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: Did you notice `duplicateCurrentToken();`? Replacing the `AddSuffixFilter` implementation above with what you have here does not produce the same results

**Comment by synhershko** [comment]
> Which is why there should be a seperation. Exact lookups should be marked by us, it makes it a lot easier to forward it down the stream rather then thinking of "why was this marked as keyword?"

Yes, but if we marked something as original token, we need to protect it down the chain as well - so mark it as keyword as well

File: java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java (hunk 1 above)
Comment (EgozyN) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: It's a different AddSuffixTokenFilter that I wrote which is much more simplified and uses HebrewTypeAttribute to determine if a word needs to be suffixed or not. 
https://github.com/EgozyN/HebMorph/blob/hebmorph-2.1/java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/AddSuffixTokenFilter.java

Comment (synhershko) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: Not my point, some tokens need to be duplicated (i.e. in IndexingAnalyzer ים is going to be indexed both as ים and ים$, at the same index poisition). Your change prevents that from happening.

Comment (EgozyN) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: That is what currently happening. As most analyzers did that by default for non-numeric tokens, the duplication is done at the lemmatizer and a filter would remove duplicates if necessary.
These regression tests are all checking for these cases:
https://github.com/EgozyN/HebMorph/commit/345e3188f589a2d7d85f55795b584ff25168799b

Comment (synhershko) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: can you please point at the line(s) in the lemmatizer that do this?

Comment (synhershko) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: also, please add a unit test around the lemmatizer itself validating that

Comment (EgozyN) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/HebrewQueryLightAnalyzer.java]: https://github.com/EgozyN/HebMorph/blob/hebmorph-2.1/java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/HebrewLemmatizerTokenFilter.java#L131
It returns true after adding the lemma to the list. This means that it adds all values to the list, and then returns the current value. Afterwards, it returns the lemmas. (return as in set the attributes). The original word is still there.
This is already validated in tests:
https://github.com/EgozyN/HebMorph/blob/hebmorph-2.1/java/src/test/java/org/apache/lucene/analysis/hebrew/TokenFilters/TestHebrewLemmatizerTokenFilter.java#L45

File: java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/HebrewLemmatizerTokenFilter.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+            // try to lemmatize
+            List<HebrewToken> tempList = lemmatizer.lemmatize(word);
+            // word wasn't found in the dictionary - try tolerating it
+            if (tempList.isEmpty()) {
+                tempList = lemmatizer.lemmatizeTolerant(word);
+                previousTolerated = true;
+            }
+            // add words to the previousLemmas : remove duplicates and tokens which aren't ranked high enough if they came from tolerating.
+            // TODO: consider the ranking as an additional filter
```
Comment (synhershko) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/HebrewLemmatizerTokenFilter.java]: it already was - why was it removed?

Comment (EgozyN) [inline java/src/main/java/org/apache/lucene/analysis/hebrew/TokenFilters/HebrewLemmatizerTokenFilter.java]: What was? Ranking? This is done at the Lemmatizer. The logic that ranks is done while lemmatizing, so making it into a another filter probably isn't the way to go.

**Pull request merged**